### ISSUE TITLE
Example of using pexpect with pyte

### ISSUE THREAD

**Issue by dbarnett** [pexpect/pexpect#587]
For curses programs, the pexpect FAQ says

> We suggest using the pyte library to screen-scrape.

Is there a simple example of how to do that, say how to start a vim instance and read out the characters on screen?

**Comment by eldipa** [comment]
I don't have a "simple" example but I can show you how I use ``pexpect`` and ``pyte`` in [byexample](https://byexamples.github.io). I hope that you have a much clearer idea of how those two libs work together after this.

First, I [create](https://github.com/byexamples/byexample/blob/251a37f3119d4c3abce124a3b4a3ab2866466e05/byexample/runner.py#L162-L166) a ``Stream`` and a ``Screen``:
```python
    def _create_terminal(self, options):
        rows, cols = options['geometry']

        self._screen = Screen(cols, rows)
        self._stream = Stream(self._screen)
```

Then, I [run](https://github.com/byexamples/byexample/blob/251a37f3119d4c3abce124a3b4a3ab2866466e05/byexample/runner.py#L90-L100)  ``pexpect.spawn`` as usual.
```python
        rows, cols = options['geometry']
        self._terminal_default_geometry = (rows, cols)

        env = os.environ.copy()
        env.update({'LINES': str(rows), 'COLUMNS': str(cols)})

        self.interpreter = pexpect.spawn(cmd, echo=False,
                                                encoding=self.encoding,
                                                dimensions=(rows, cols),
                                                env=env)
```

In both cases I configure the same terminal dimension (rows and columns). I found that some programs honor the environment variables ``LINES`` and ``COLUMNS`` while others honor the 
configuration of ``pexpect`` (If you track this you will see that it uses magical ``TIOCSWINSZ`` setting [here](https://github.com/pexpect/ptyprocess/blob/3931cd45db50ee8533b8b0fef424b8d75f7ba1c2/ptyprocess/ptyprocess.py#L118-L126))

Now, you [send data](https://github.com/byexamples/byexample/blob/251a37f3119d4c3abce124a3b4a3ab2866466e05/byexample/runner.py#L151)  to the process as usual con ``pexpect``:
```python
            self.interpreter.sendline(line)
```

And you can [read the raw data](https://github.com/byexamples/byexample/blob/251a37f3119d4c3abce124a3b4a3ab2866466e05/byexample/runner.py#L252-L256) as well:
```python
        expect = [prompt_re, pexpect.TIMEOUT]
        PS_found, Timeout = range(len(expect))

        what = self.interpreter.expect(expect, timeout=timeout)
        self.last_output.append(self.interpreter.before)
```

What you read from ``pexpect`` is the *raw output* from the process, possibly full of *control sequences* to be interpreted by ``pyte``.

So we grab all this output from ``pexpect`` and we pass it to ``pyte``, then you can get the lines of the console from ``pyte`` *after* the processing (and remotion) of the control sequences:
```python
            out = self._emulate_ansi_terminal(self.last_output)
```
See this [part 1](https://github.com/byexamples/byexample/blob/251a37f3119d4c3abce124a3b4a3ab2866466e05/byexample/runner.py#L269)

```python
    def _emulate_ansi_terminal(self, chunks, join=True):
        for chunk in chunks:
            self._stream.feed(chunk)

        lines = self._screen.display
        self._screen.reset()
        if str == bytes:
            # Python 2.7 support only: it works on str/bytes only
            # XXX this is a limitation, if the output has a single non-ascii
            # character this will blow up without the 'ignore'
            lines = (str(line.rstrip().encode('ascii', 'ignore')) for line in lines)
        else:
            lines = (line.rstrip() for line in lines)

        return '\n'.join(lines) if join else lines
```
See this [part 2](https://github.com/byexamples/byexample/blob/251a37f3119d4c3abce124a3b4a3ab2866466e05/byexample/runner.py#L207-L221)

You will see that each line from ``pyte`` has a fixed amount of columns, filled with whitespace at the right. In my case I was not interested in that so I did a ``rstrip`` but it is optional.
Also, once I got the lines from ``pyte`` I reset it with ``self._screen.reset()`` so in the next call to ``_emulate_ansi_terminal`` I don't get repeated lines.
This again will depend of what you want to do. 

I use ``pexpect`` and ``pyte`` to support [ANSI Terminals](https://byexamples.github.io/byexample/advanced/terminal-emulation) during the execution of the tests. I should work both in Python 2 and 3.

I hope that this serves as a starting point, let me know if you need an extra hand.

**Comment by Red-M** [comment]
I'm going to track adding this as an example, I'd like it to be there for numerous reasons, one of them being controlling tmux session due to tmux using ncurses.

**Comment by eldipa** [comment]
@Red-M ask me anything that you need if you are in trouble. I was thinking in adding an example too so I'm more than willing to help.

**Comment by Red-M** [comment]
So I've been through this and all you need to do is feed a pyte stream the data from the terminal (or in this case your process) and it will generate a view of that in the display property of the pyte screen associated to the pyte stream.

**Comment by eldipa** [comment]
That's correct: the integration between `pexpect` and `pyte` is very simple. There are only a few details that one should be aware of like resetting the screen, stripping unwanted whitespace and setting the size of the terminal.

I think that those are the interesting bits that should be in the example. Believe me, it took me a while to figure out some of them!
